RavenDB-19748: New categories in Traffic Watch

TheGoldenPlatypus opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19748/New-categories-in-Traffic-Watch

### Additional description

Added missing TrafficWatchChangeType & CustomInfo values to be displayed on Traffic Watch screen

### Type of change

- Optimization

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- Yes. v5.4

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- It has been verified by manual testing

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No
- 
### UI work

- It requires further work in the Studio : new TrafficWatchChangeType categories should be added to drop down menu

TheGoldenPlatypus committed:
RavenDB-19748: fix missing type&customInfo values at traffic watch

TheGoldenPlatypus committed:
RavenDB-19598: added new types to TrafficWatchChangeType enum

ppekrol commented:
Something went wrong here, I see
- RavenDB-19748 which is a correct issue
- RavenDB-19598 which is a incorrect one

PR Title: `Ravendb v5.4 19748 060223` which does not match any of my regex expressions :)

Issue link to https://issues.hibernatingrhinos.com/issue/RavenDB-19598/Add-new-Refresh-overloads which does not match what was done here

Can we fix those first?

TheGoldenPlatypus commented:
> Something went wrong here, I see
> 
> * RavenDB-19748 which is a correct issue
> * RavenDB-19598 which is a incorrect one
> 
> PR Title: `Ravendb v5.4 19748 060223` which does not match any of my regex expressions :)
> 
> Issue link to https://issues.hibernatingrhinos.com/issue/RavenDB-19598/Add-new-Refresh-overloads which does not match what was done here
> 
> Can we fix those first?

changed the PR title, and I replaced the issue Link

garayx commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -103,6 +108,9 @@ public async Task WaitForIndex()
             await ServerStore.Cluster.WaitForIndexNotification(index);
+            
+            if (TrafficWatchManager.HasRegisteredClients)
+                AddStringToHttpContext("WaitForIndexCommand", TrafficWatchChangeType.ClusterCommands);
```
can we use `nameof()` ?

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -121,7 +129,10 @@ public Task SuspendObserver()
             RedirectToLeader();
-
+            
+            if (TrafficWatchManager.HasRegisteredClients)
```
this should go inside the `if (ServerStore.IsLeader())` clause

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -121,7 +129,10 @@ public Task SuspendObserver()
-
+            
+            if (TrafficWatchManager.HasRegisteredClients)
+                AddStringToHttpContext("SuspendObserverCommand", TrafficWatchChangeType.ClusterCommands);
```
can we use nameof() ?

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -287,6 +315,9 @@ public async Task Bootstrap()
             NoContentStatus();
+
+            if (TrafficWatchManager.HasRegisteredClients)
+                AddStringToHttpContext("BootstrapCommand", TrafficWatchChangeType.ClusterCommands);
```
can we use nameof() ?

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -509,6 +540,9 @@ public async Task AddNode()
             RedirectToLeader();
+            
+            if (TrafficWatchManager.HasRegisteredClients)
+                AddStringToHttpContext("AddNodeCommand", TrafficWatchChangeType.ClusterCommands);
```
can we use nameof() ?

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -557,6 +591,9 @@ public async Task DeleteNode()
             RedirectToLeader();
+           
+            if (TrafficWatchManager.HasRegisteredClients)
+                AddStringToHttpContext("DeleteNodeCommand", TrafficWatchChangeType.ClusterCommands);
```
can we use nameof() ?

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -569,19 +606,29 @@ public async Task SetLicenseLimit()
             NoContentStatus();
+
+            if (TrafficWatchManager.HasRegisteredClients)
+                AddStringToHttpContext("SetLimitCommand", TrafficWatchChangeType.ClusterCommands);
```
can we use nameof() ?

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -557,6 +591,9 @@ public async Task DeleteNode()
 
             RedirectToLeader();
+           
+            if (TrafficWatchManager.HasRegisteredClients)
```
this should be inside the if?

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -569,19 +606,29 @@ public async Task SetLicenseLimit()
             NoContentStatus();
+
+            if (TrafficWatchManager.HasRegisteredClients)
+                AddStringToHttpContext("TimeoutCommand", TrafficWatchChangeType.ClusterCommands);
```
```suggestion
                AddStringToHttpContext(nameof(TimeoutNow), TrafficWatchChangeType.ClusterCommands);
```

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -569,19 +606,29 @@ public async Task SetLicenseLimit()
         public Task EnforceReelection()
         {
+            if (TrafficWatchManager.HasRegisteredClients)
+                AddStringToHttpContext("ReelectCommand", TrafficWatchChangeType.ClusterCommands);  
```
nameof

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/NotificationCenter/Handlers/ServerNotificationCenterHandler.cs:
```diff
@@ -47,8 +49,10 @@ public async Task Get()
                     await writer.WriteNotifications(isValidFor);
+
+                    if (TrafficWatchManager.HasRegisteredClients)
+                        AddStringToHttpContext(writer.ToString(), TrafficWatchChangeType.ClusterCommands);
```
should be `TrafficWatchChangeType.Notifications` ?

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/Web/Operations/OperationsHandler.cs:
```diff
@@ -29,6 +29,10 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
                     writer.WritePropertyName(nameof(GetNextOperationIdCommand.NodeTag));
                     writer.WriteString(Server.ServerStore.NodeTag);
                     writer.WriteEndObject();
+
```
we have to the addition of Operations (`Database.Operations.AddOperation()`) as well.

TheGoldenPlatypus replied:
Done.
Notice: the  requestHandler propperty marked as [CanBeNull] because of Migrator.StartMigratingSingleDatabase() & BasicOperationsTests which are not inherit any handler

garayx commented on src/Raven.Server/Documents/PeriodicBackup/PeriodicBackupRunner.cs:
```diff
@@ -429,6 +429,7 @@ private long CreateBackupTask(PeriodicBackup periodicBackup, bool isFullBackup,
                         Operations.Operations.OperationType.DatabaseBackup,
                         taskFactory: onProgress => StartBackupThread(periodicBackup, backupTask, tcs, onProgress),
                         id: operationId,
+                        null,
```
```suggestion
                        requestHandler: null,
```

garayx commented on src/Raven.Server/NotificationCenter/Handlers/ServerNotificationCenterHandler.cs:
```diff
@@ -47,8 +49,10 @@ public async Task Get()
                     writer.AfterTrackActionsRegistration = ServerStore.NotifyAboutClusterTopologyAndConnectivityChanges;
-
                     await writer.WriteNotifications(isValidFor);
+
```
where do we log the adding of notifications ?

TheGoldenPlatypus committed:
RavenDB-19748: addressed pr comments

TheGoldenPlatypus committed:
RavenDB-19748: removed most previous additions

garayx commented on src/Raven.Server/Documents/Handlers/DocumentHandler.cs:
```diff
@@ -100,6 +103,9 @@ public async Task GetDocSize()
                 {
                     context.Write(writer, documentSizeDetails.ToJson());
+                    
+                    if (TrafficWatchManager.HasRegisteredClients)
```
we are returning earlier in this method on `if (document == null)`, and we still need to log

I wondering if we can do:
```
if (TrafficWatchManager.HasRegisteredClients)
   AddStringToHttpContext("N/A", TrafficWatchChangeType.Documents);
```

In the constructor of this class so in case of exception we still will log in under `Documents` category

TheGoldenPlatypus replied:
Done

TheGoldenPlatypus committed:
RavenDB-19748: added default log lone to DocumentHandler constructor

garayx commented on src/Raven.Server/Documents/Handlers/DocumentHandler.cs:
```diff
@@ -47,12 +47,22 @@ namespace Raven.Server.Documents.Handlers
             var id = GetQueryStringValueAndAssertIfSingleAndNotEmpty("id");
             var changeVector = GetStringFromHeaders(Constants.Headers.IfNoneMatch);
 
+            if (TrafficWatchManager.HasRegisteredClients)
```
we can remove this since we log in the ctor now

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/NotificationCenter/Handlers/ServerNotificationCenterHandler.cs:
```diff
@@ -33,7 +35,9 @@ public async Task Get()
-
+                            
+                            if (TrafficWatchManager.HasRegisteredClients)
+                                AddStringToHttpContext(writer.ToString(), TrafficWatchChangeType.Notifications);
```
we need to write the `action.ToJson()` here since we reset the ctx in `writer.WriteToWebSocket()`

TheGoldenPlatypus replied:
Done

garayx commented on src/Raven.Server/NotificationCenter/Handlers/ServerNotificationCenterHandler.cs:
```diff
@@ -42,12 +46,13 @@ public async Task Get()
                         var action = OperationChanged.Create(null, operation.Id, operation.Description, operation.State, operation.Killable);
 
+                        if (TrafficWatchManager.HasRegisteredClients)
+                            AddStringToHttpContext(writer.ToString(), TrafficWatchChangeType.Notifications);
```
we need to write the `action.ToJson()` here since we reset the ctx in `writer.WriteToWebSocket()`

TheGoldenPlatypus replied:
Done

TheGoldenPlatypus committed:
RavenDB_19748: removed redundant lines, addressed pr comments

garayx commented on src/Raven.Server/Documents/Handlers/DocumentHandler.cs:
```diff
@@ -473,6 +483,8 @@ public async Task Delete()
                 var cmd = new DeleteDocumentCommand(id, changeVector, Database);
                 await Database.TxMerger.Enqueue(cmd);
+                if (TrafficWatchManager.HasRegisteredClients)
+                    AddStringToHttpContext(id, TrafficWatchChangeType.Documents);
```
we can remove this since we log in the ctor now

TheGoldenPlatypus replied:
Done

TheGoldenPlatypus committed:
RavenDB_19748: removed redundant log line
